trie/trie.py

```diff
@@ -378,3 +397,256 @@ def __delitem__(self, key):
+            if len(left_child) == common_prefix_len + 1:
+                oldnode = right_child
+            # (ii), (iv), (vi), (viii)
+            else:
+                oldnode = self._hash_and_save(
+                    encode_kv_node(left_child[common_prefix_len+1:], right_child)
+                )
+            # Create the new branch node (because the key paths diverge, there has to
+            # be some "first bit" at which they diverge, so there must be a branch
+            # node somewhere)
+            if keypath[common_prefix_len:common_prefix_len+1] == BYTE_1:
+                newsub = self._hash_and_save(encode_branch_node(oldnode, valnode))
+            else:
+                newsub = self._hash_and_save(encode_branch_node(valnode, oldnode))
+            # Case 1: keypath prefixes match in the first bit, so we still need
+            # a kv node at the top
+            # (i) (ii) (iii) (iv)
+            if common_prefix_len:
+                return self._hash_and_save(
+                    encode_kv_node(left_child[:common_prefix_len], newsub)
+                )
+            # Case 2: keypath prefixes diverge in the first bit, so we replace the
+            # kv node with a branch node
+            # (v) (vi) (vii) (viii)
+            else:
+                return newsub
+
+    def _set_branch_node(self, keypath, node_type, left_child, right_child, value):
+        new_left_child, new_right_child = left_child, right_child
+        # Which child node to update? Depends on first bit in keypath
+        if keypath[:1] == BYTE_0:
+            new_left_child = self._set(left_child, keypath[1:], value)
+        else:
+            new_right_child = self._set(right_child, keypath[1:], value)
+        # Compress branch node into kv node
+        if new_left_child == BLANK_HASH or new_right_child == BLANK_HASH:
+            subnodetype, sub_left_child, sub_right_child = parse_node(
+                self.db[
+                    new_left_child
+                    if new_left_child != BLANK_HASH
+                    else new_right_child]
+            )
+            first_bit = BYTE_1 if new_right_child != BLANK_HASH else BYTE_0
+            # Compress (k1, (k2, NODE)) -> (k1 + k2, NODE)
+            if subnodetype == KV_TYPE:
+                return self._hash_and_save(
+                    encode_kv_node(first_bit + sub_left_child, sub_right_child)
+                )
+            # kv node pointing to a branch node
+            elif subnodetype in (BRANCH_TYPE, LEAF_TYPE):
+                return self._hash_and_save(
+                    encode_kv_node(
+                        first_bit,
+                        new_left_child
+                        if new_left_child != BLANK_HASH
+                        else new_right_child
+                    )
+                )
+        else:
+            return self._hash_and_save(encode_branch_node(new_left_child, new_right_child))
+
+    def exists(self, key):
+        validate_is_bytes(key)
```

Comment:
You should be able to remove this validation check since it occurs within `self.get(...)`